src/Castle.ts

```diff
@@ -0,0 +1,218 @@
+import fetch from 'node-fetch';
+import { reduce } from 'lodash';
+import { IncomingHttpHeaders } from 'http2';
+import AbortController from 'abort-controller';
+
+const defaultApiUrl = 'https://api.castle.io';
+
+type TrackParameters = {
+  event: string;
+  user_id: string;
+  user_traits?: object;
+  created_at?: string;
+  context: {
+    ip: string;
+    client_id: string;
+    headers: IncomingHttpHeaders;
+  };
+};
+
+type AuthenticateResult = {
+  action: string;
+  user_id?: string;
+  device_token?: string;
+  failover?: boolean;
+  failover_reason?: string;
+};
+
+const isTimeout = (e: Error) => e.name === 'AbortError';
+
+export class Castle {
+  private apiSecret: string;
+  private apiUrl: string;
+  private timeout: number;
+  private allowedHeaders: string[];
+  private disallowedHeaders: string[];
+  private overrideFetch: any;
```

Comment:
what is overrideFetch used for ?

Reply:
For testing, and technically (doubt they ever would) if implementer wants to use their own fetch.